ibis/backends/polars/compiler.py

```diff
@@ -1167,10 +1167,37 @@ def execute_count_distinct_star(op, **kw):
     return arg.n_unique()
 
 
+_UDF_INVOKERS = {
+    # Convert polars series into a list
+    #   -> map the function element by element
+    #   -> convert back to a polars series
+    InputType.PYTHON: lambda func, dtype, args: pl.Series(
+        map(func, *(arg.to_list() for arg in args)),
+        dtype=dtype_to_polars(dtype),
+    ),
+    # Convert polars series into a pyarrow array
+    #  -> invoke the function on the pyarrow array
+    #  -> cast the result to match the ibis dtype
+    #  -> convert back to a polars series
+    InputType.PYARROW: lambda func, dtype, args: pl.from_arrow(
+        func(*(arg.to_arrow() for arg in args)).cast(dtype.to_pyarrow()),
+    ),
+}
+
+
 @translate.register(ops.ScalarUDF)
 def execute_scalar_udf(op, **kw):
-    if op.__input_type__ == InputType.BUILTIN:
+    if (input_type := op.__input_type__) in (InputType.PYARROW, InputType.PYTHON):
```

Comment:
Two nits:
- Can use `_UDF_INVOKERS` instead of a tuple for membership checking here
- Since `input_type` is used in every branch below, it feels cleaner to me to define it at the top-level rather than using `:=`. IMO `:=` is clearest when the new variable is only used in the block below the statement (e.g. within the corresponding `if`/`for` block).

```suggestion
    input_type = op.__input_type__
    if input_type in _UDF_INVOKERS:
```